### PR TITLE
[WIP] Issue 29: HTTP API Support

### PR DESCRIPTION
**Pull request by Lull3rSkat3r**
(c) Copyright IBM Corp. 2015

### CODE REVIEW

File: kernel_gateway/gatewayapp.py [hunk 1]
```diff
@@ -144,9 +146,23 @@ def _default_kernel_name_default(self):
...
+    )
+    def _list_kernels_default(self):
+        return os.getenv(self.list_kernels_env, 'False') == 'True'
+
+    api_env = 'KG_API'
+    api = Unicode(config=True,
+        help='Designates that the kernel gateway will serve APIs from this notebook (KG_API env var)'
+    )
+    def _api_default(self):
+        return os.getenv(self.api_env)
```
Comment (parente) [inline kernel_gateway/gatewayapp.py]: Probably want this returning a default string representing the "notebook kernel API + jupyter protocol over websocket" option. Why: we've got one protocol+api, you guys are working on a second (notebook defined API over HTTP), and there might be more coming down the pipe (e.g., see #17). Best to plan for extensibility here.

Make the help print the available valid options. You can get it to show the default too: see the others where I set the value + used the default method.

File: kernel_gateway/services/notebooks/handlers.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+from .request_utils import *
+
+class NotebookAPIHandler(tornado.web.RequestHandler):
+    get_source = None
+    put_source = None
+    post_source = None
+    delete_source = None
+    kernel_client = None
+    execution_timeout = 5
+    def initialize(self, sources, kernel_client):
```
Comment (parente) [inline kernel_gateway/services/notebooks/handlers.py]: IIRC `initialize` is invoked per instance of this class in tornado, and *every web request spawns its own instance*. That's why the handler map is to classes, not instances.

Definitely don't want to be doing this parsing over and over again. Best to keep these out of the class at the module level (singleton) or as class vars.

File: kernel_gateway/gatewayapp.py [hunk 3]
```diff
@@ -213,17 +258,47 @@ def init_webapp(self):
...
-            default_kernelspec_handlers +
-            default_base_handlers
-        ):
-            # Create a new handler pattern rooted at the base_url
-            pattern = url_path_join(self.base_url, handler[0])
-            # Some handlers take args, so retain those in addition to the
-            # handler class ref
-            new_handler = tuple([pattern] + list(handler[1:]))
-            handlers.append(new_handler)
+        if self.api:
```
Comment (parente) [inline kernel_gateway/gatewayapp.py]: I know it's early and you haven't refactored yet, but I think the logic below should ultimately wind up in a function or functions defined in the `notebooks/handlers` module to encapsulate the parsing logic + execution logic in one spot? Invoke the function from here, but define it there.

Compelling reason to do it: I think these "handlers" are going to become part of a kind of lite plugin system. As long as you write a Python package/module with such-and-such API, you can plug it into kernel gateway to define new ways of talking to kernels.

Comment (nitind) [inline kernel_gateway/gatewayapp.py]: We're definitely going to do some refactoring--the kernel spec resides in the notebook, but we're currently set to only read that after the KernelManager is already created.

File: kernel_gateway/services/notebooks/handlers.py (hunk 2 above)
Comment (nitind) [inline kernel_gateway/services/notebooks/handlers.py]: At the moment, the init_webapp function of gatewayapp.py is sending the parsed cell source and URI to Tornado when registering the hander class, so however messy it is right now, we're only parsing things once.

**Pull request closed without merging**